### ISSUE TITLE
ignored __stdcall

### ISSUE THREAD

**Issue by Dushistov** [servo/rust-bindgen#547]
This simplified example to demonstrate error, that cause random crashes,
in jni dll (windows 7 32bit) written on rust. The cause is silent replace of `__stdcall` calling conversation
with `__cdecl` calling conversation:

### Input C/C++ Header

```C++
struct JNINativeInterface_ {
  int (__stdcall *GetVersion)(void *env);
};
```

### Bindgen Invokation

```
$ bindgen test.h 
```

### Actual Results

```rust
#[repr(C)]
#[derive(Debug, Copy)]
pub struct JNINativeInterface_ {
    pub GetVersion: ::std::option::Option<unsafe extern "C" fn(env:
                                                                   *mut ::std::os::raw::c_void)
                                              -> ::std::os::raw::c_int>,
}
```
### Expected Results

```rust
#[repr(C)]
#[derive(Debug, Copy)]
pub struct JNINativeInterface_ {
    pub GetVersion: ::std::option::Option<unsafe extern "stdcall" fn(env:
                                                                   *mut ::std::os::raw::c_void)
                                              -> ::std::os::raw::c_int>,
}
```

**Comment by emilio** [comment]
Huh, we should keep track of that, thanks for the report.

**Comment by emilio** [comment]
So, fun part is I can't reproduce because I don't have a windows machine. That being said, I can probably find something or coerce libclang to show it.

Meanwhile, if you could do the following for me, it'd be super-helpful:

 * Run bindgen with:

```c
__stdcall void bar();
```

 * Run your example with `--emit-clang-ast`.
 * Run your example with `RUST_LOG=bindgen`

And post the results here, it'd be great :)

Thanks in advance!

**Comment by emilio** [comment]
See https://github.com/servo/rust-bindgen/pull/549, this is arguably a libclang bug :(. That patch works around the issue and should work.

Thanks for reporting!

**Comment by Dushistov** [comment]
here result of

```
set RUST_LOG=bindgen
bindgen --version
INFO:bindgen: Clang Version: clang version 3.9.0 (branches/release_39)
bindgen 0.22.0
bindgen --emit-clang-ast test3.h > clang-output.txt 2>rust-log.txt


```[clang-output.txt](https://github.com/servo/rust-bindgen/files/809684/clang-output.txt)
[rust-log.txt](https://github.com/servo/rust-bindgen/files/809685/rust-log.txt)

**Comment by Dushistov** [comment]
> So, fun part is I can't reproduce because I don't have a windows machine. 

I used virtual box image, that provide Microsoft for free:

https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/

>See #549, this is arguably a libclang bug :(. That patch works around the issue and should work.

note, that by default bindgen emit code with `derive(Debug)`, but this is actually not works
for stdcall fn pointers: see [my rustc bug report](https://github.com/rust-lang/rust/issues/40158)

**Comment by emilio** [comment]
> note, that by default bindgen emit code with derive(Debug), but this is actually not works
for stdcall fn pointers.

😞 

I'll have to add an explicit path for that I guess. Sigh. Does `derive(Copy)` work?

**Comment by Dushistov** [comment]
>I'll have to add an explicit path for that I guess. Sigh. 

For note: this is Sun/Oracle guys fault, to use `__stdcall` for JNI interface for Java, not my.

> Does derive(Copy) work?

Yes

**Comment by emilio** [comment]
> For note: this is Sun/Oracle guys fault in usage of __stdcall for JNI interface for Java, not my :)

Oh, sure, sorry if I made you think this was somehow your fault. It's just quite disappointing to see this intersection between libclang and rustc bugs that makes bindgen even more complicated than what it is :(

**Comment by emilio** [comment]
Actually reading that bug and the rust source seems even funnier. Since there's a hard-coded limit in the number of arguments they implement these traits for, that means that bindgen will not produce compilable code for stuff like:

```
typedef void (*my_fun_t)(int, int, int, int, int, int, int, int, int, int, int,
                         int, int, int, int, int);

struct Foo {
  my_fun_t callback;
};
```

**Comment by emilio** [comment]
I added a fix for that too at #549.

**Comment by Dushistov** [comment]
Thanks, I tested on windows 7 (32 bit), and at now my project(https://github.com/Dushistov/rust_swig) works on windows without problems, at least all tests passed.
